src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs

```diff
@@ -208,25 +208,56 @@ static uint GetTempPathW(int bufferLen, ref char buffer)
         // name on disk.
         public static string GetTempFileName()
         {
-            var tempPathBuilder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            GetTempPath(ref tempPathBuilder);
-
-            var builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            uint result = Interop.Kernel32.GetTempFileNameW(
-                ref tempPathBuilder.GetPinnableReference(), "tmp", 0, ref builder.GetPinnableReference());
-
-            tempPathBuilder.Dispose();
-
-            if (result == 0)
-                throw Win32Marshal.GetExceptionForLastWin32Error();
-
-            builder.Length = builder.RawChars.IndexOf('\0');
+            // Avoid GetTempFileNameW because it is limited to 0xFFFF possibilities, which both
+            // means that it may have to make many attempts to create the file before
+            // finding an unused name, and also that if an app "leaks" such temp files,
+            // it can prevent GetTempFileNameW succeeding at all.
+            //
+            // To make this a little more robust, generate our own name with more
+            // entropy. We could use GetRandomFileName() here, but for consistency
+            // with Unix and to retain the ".tmp" extension we will use the "tmpXXXXXX.tmp" pattern.
+            // Using 32 characters for convenience, that gives us 32^^6 ~= 10^^9 possibilities,
+            // but we'll still loop to handle the unlikely case the file already exists.
+
+            byte* bytes = stackalloc byte[KeyLength];
+            Span<char> span = stackalloc char[13]; // tmpXXXXXX.tmp
+
+            int i = 0;
+            while (true)
+            {
+                const int KeyLength = 4; // 4 bytes = more than 6 x 5 bits
+                Interop.GetRandomBytes(bytes, KeyLength);
+
+                byte b0 = bytes[0];
+                byte b1 = bytes[1];
+                byte b2 = bytes[2];
+                byte b3 = bytes[3];
+
+                span[0] = span[10] = 't';
+                span[1] = span[11] = 'm';
+                span[2] = span[12] = 'p';
+                span[3] = (char)Base32Char[b0 & 0b0001_1111];
+                span[4] = (char)Base32Char[b1 & 0b0001_1111];
+                span[5] = (char)Base32Char[b2 & 0b0001_1111];
+                span[6] = (char)Base32Char[b3 & 0b0001_1111];
+                span[7] = (char)Base32Char[((b0 & 0b1110_0000) >> 5) | ((b1 & 0b1100_0000) >> 3)];
+                span[8] = (char)Base32Char[((b2 & 0b1110_0000) >> 5) | ((b3 & 0b1100_0000) >> 3)];
+                span[9] = '.';
+
+                string path = string.Concat(Path.GetTempPath(), span);
+
+                try
+                {
+                    File.OpenHandle(path, FileMode.CreateNew, FileAccess.Write).Dispose();
```

Comment:
Do we need to ensure the Temp directory exists before calling this? I'm not sure if GetTempFileNameW did that for us or not.

Reply:
Looking at the code -- GetTempPathW() does not check the disk. Eg., if you set %TMP%, Windows will return that value, whether or not it's even a valid path. Generally, of course, %TMP% and %TEMP% aren't set, so it will return the default location, which always (?) exists. 

However GetTempFileNameW() does first check that what it gets from GetTempPathW() exists, and immediately returns ERROR_DIRECTORY if it doesn't. 

So good catch. I think what I should do is be optimistic, then loop only if I get AccessDeniedException and also probably File.Exists.